src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,701 @@
+//! Script which uses `cargo test -Z build-std` to generate or execute
+//! `mir-json` commands that build the modified standard libraries.
+
+#![feature(rustc_private)]
+
+extern crate cargo_metadata;
+extern crate rustc_driver;
+extern crate rustc_session;
+extern crate serde;
+extern crate shell_escape;
+extern crate tempfile;
+
+use std::{
+    collections::HashMap,
+    convert::TryInto,
+    env, fmt, fs,
+    io::BufReader,
+    os::unix,
+    process::{Command, Stdio},
+};
+
+use cargo_metadata::{
+    camino::{Utf8Path, Utf8PathBuf},
+    Edition, Message, PackageId, Target,
+};
+use serde::Deserialize;
+use shell_escape::escape;
+use tempfile::{tempdir, TempDir};
+
+mod extra_libs {
+    pub const CRUCIBLE: &str = "crucible";
+    pub const INT512: &str = "int512";
+    pub const BYTES: &str = "bytes";
+    pub const BYTEORDER: &str = "byteorder";
+}
+
+/// Name of the new empty cargo project to be created to run `cargo test -Z
+/// build-std` in.
+const EMPTY_PROJECT_NAME: &str = "mir-json-translate-libs-empty-project";
+
+/// Deserialized form of cargo --unit-graph JSON output.
+/// See https://doc.rust-lang.org/cargo/reference/unstable.html#unit-graph
+#[derive(Deserialize)]
+struct UnitGraph {
+    version: u32,
+    units: Vec<UnitGraphUnit>,
+    roots: Vec<usize>,
+}
+
+#[derive(Deserialize)]
+struct UnitGraphUnit {
+    pkg_id: PackageId,
+    target: Target,
+    dependencies: Vec<UnitGraphDependency>,
+}
+
+#[derive(Clone, Deserialize)]
+struct UnitGraphDependency {
+    index: usize,
+    extern_crate_name: CrateName,
+}
+
+/// Module to hide field of newtype [CrateName].
+mod crate_name {
```

Comment:
Enforcing the invariant this way is probably a good idea, but honestly, I normally don't bother.  For something that's used only in a fairly small scope (just this file), I don't think it's too bad to rely on people either reading the docs before they use it or copying how it's used in existing examples.  Though I will still use `pub` on fields/methods of private types just as a hint that "other code in the module should generally access these things, but not those ones".

Reply:
Ok. I'll leave this one this way since I already did it, but will keep that in mind for the future!